libraries/Microsoft.Bot.Builder.Dialogs/DialogContextState.cs

```diff
@@ -218,6 +230,11 @@ public void SetValue(string pathExpression, object value)
                 throw new Exception($"{pathExpression} = You can't pass an unresolved Task to SetValue");
             }
 
+            if (pathExpression.StartsWith("$"))
```

Comment:
I call a resolvePath helper that expands not only $ but # and @ as well.  Also you're only expanding $ for SetValue(). This needs be done for the others like GetValue(), and RemoveValue().